File: Python-DSA-Ejemplos/data_structures/linked_list.py
"""

Implementación de una Lista Enlazada (Singly Linked List).

Una lista enlazada es una estructura de datos lineal donde los elementos no se almacenan
en ubicaciones de memoria contiguas. Consiste en nodos, donde cada nodo contiene
un dato y una referencia (o puntero) al siguiente nodo en la secuencia.
"""

class Node:
    """Un nodo individual de una lista enlazada."""
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    """Clase que representa la lista enlazada completa."""
    def __init__(self):
        """Inicializa una lista vacía."""
        self.head = None

    def is_empty(self):
        """Comprueba si la lista está vacía."""
        return self.head is None

    def append(self, data):
        """Añade un nodo al final de la lista."""
        new_node = Node(data)
        if self.is_empty():
            self.head = new_node
            return
        
        last_node = self.head
        while last_node.next:
            last_node = last_node.next
        last_node.next = new_node

    def prepend(self, data):
        """Añade un nodo al principio de la lista."""
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def delete(self, data):
        """Elimina la primera ocurrencia de un nodo con el dato especificado."""
        if self.is_empty():
            return

        if self.head.data == data:
            self.head = self.head.next
            return

        current_node = self.head
        while current_node.next and current_node.next.data != data:
            current_node = current_node.next
        
        if current_node.next:
            current_node.next = current_node.next.next
        else:
            print(f"Dato '{data}' no encontrado en la lista.")

    def display(self):
        """Muestra los elementos de la lista."""
        elements = []
        current_node = self.head
        while current_node:
            elements.append(str(current_node.data))
            current_node = current_node.next
        print(" -> ".join(elements) + " -> None")

# --- Ejemplo de Uso ---
if __name__ == "__main__":
    lista = LinkedList()
    print(f"¿Lista vacía? {lista.is_empty()}")

    print("\n--- Añadiendo elementos (append) ---")
    lista.append("A")
    lista.append("B")
    lista.append("C")
    lista.display()

    print("\n--- Añadiendo al inicio (prepend) ---")
    lista.prepend("Inicio")
    lista.display()
    print(f"¿Lista vacía? {lista.is_empty()}")

    print("\n--- Eliminando elementos ---")
    lista.delete("B")
    print("Después de eliminar 'B':")
    lista.display()

    lista.delete("Inicio")
    print("Después de eliminar 'Inicio':")
    lista.display()

    lista.delete("Z") # Intentar eliminar un elemento que no existe

File: Go-DSA-Ejemplos/data_structures/linked_list.go
package data_structures

import "fmt"

type Node struct {
	data string
	next *Node
}

type LinkedList struct {
	head *Node
}

func (ll *LinkedList) IsEmpty() bool {
	return ll.head == nil
}

func (ll *LinkedList) Append(data string) {
	newNode := &Node{data: data, next: nil}
	if ll.IsEmpty() {
		ll.head = newNode
		return
	}

	lastNode := ll.head
	for lastNode.next != nil {
		lastNode = lastNode.next
	}
	lastNode.next = newNode
}

func (ll *LinkedList) Prepend(data string) {
	newNode := &Node{data: data, next: ll.head}
	ll.head = newNode
}

func (ll *LinkedList) Delete(data string) {
	if ll.IsEmpty() {
		return
	}

	// If the node to delete is the head
	if ll.head.data == data {
		ll.head = ll.head.next
		return
	}

	currentNode := ll.head
	for currentNode.next != nil && currentNode.next.data != data {
		currentNode = currentNode.next
	}

	// If the node to delete is found
	if currentNode.next != nil {
		currentNode.next = currentNode.next.next
	} else {
		fmt.Printf("Dato '%s' no encontrado en la lista.\n", data)
	}
}

func (ll *LinkedList) Display() {
	if ll.IsEmpty() {
		fmt.Println("Lista vacía.")
		return
	}
	currentNode := ll.head
	for currentNode != nil {
		fmt.Printf("%s -> ", currentNode.data)
		currentNode = currentNode.next
	}
	fmt.Println("None")
}

func LinkedListDemo() {
	lista := LinkedList{}
	fmt.Println("¿Lista vacía?", lista.IsEmpty())

	fmt.Println("\n--- Añadiendo elementos (Append) ---")
	lista.Append("A")
	lista.Append("B")
	lista.Append("C")
	lista.Display()

	fmt.Println("\n--- Añadiendo al inicio (Prepend) ---")
	lista.Prepend("Inicio")
	lista.Display()
	fmt.Println("¿Lista vacía?", lista.IsEmpty())

	fmt.Println("\n--- Eliminando elementos ---")
	lista.Delete("B")
	fmt.Println("Después de eliminar 'B':")
	lista.Display()

	lista.Delete("Inicio")
	fmt.Println("Después de eliminar 'Inicio':")
	lista.Display()

	lista.Delete("Z") // Intentar eliminar un elemento que no existe
	lista.Display()
}

func main() {
	LinkedListDemo()
}
